internal/storage/bucket_handle.go

```diff
@@ -430,6 +435,35 @@ func (b *bucketHandle) ComposeObjects(ctx context.Context, req *gcs.ComposeObjec
 	return
 }
 
+// Fetch bucket type from GetStorageLayout and set it. e.g. Hierarchical or NonHierarchical
+func (b *bucketHandle) FetchAndSetBucketType() {
+	if b.controlClient == nil {
+		b.bucketType = gcs.NonHierarchical
+		return
+	}
+
+	var callOptions []gax.CallOption
+	stoargeLayout, err := b.controlClient.GetStorageLayout(context.Background(), &controlpb.GetStorageLayoutRequest{
+		Name:      "projects/_/buckets/" + b.bucketName + "/storageLayout",
+		Prefix:    "",
+		RequestId: "",
+	}, callOptions...)
+
+	// In case bucket does not exist, set type unknown instead of panic.
```

Comment:
What happens if error comes because of some permission issue?

Reply:
`storage.objects.list` permission is require for this API call. This permission is included in `Storage Object Viewer ` role. Which is I think minimum role require to mount the bucket.

If user does not have `Storage Object Viewer `  role then anyway it will fail on very first List call.

Reply:
@raj-prince in that scenario also, bucket type should be set as unknown, and it will be treated as flat bucket

Reply:
@Tulsishah "If user does not have Storage Object Viewer  role then anyway it will fail on very first List call" --> as I can see in this PR, set type getting called before first list call inside setup bucket ? are we doing list call before setup bucket as well ?